### PR TITLE
Add Optuna optimization of model's hyperparameters

### PR DESCRIPTION
**Pull request by luca-bottero**
See https://optuna.readthedocs.io/en/stable/ for details

### CODE REVIEW

**Comment by fcatalan92** [comment]
Hi @luca-bottero, thank you very much for this contribution. I'm my opinion, optuna is a very interesting option for our package, since currently the Bayesian optimisation is performed by https://github.com/fmfn/BayesianOptimization which is not actively developed/maintained.

The CI is currently broken since you have to add the dependency to optuna in our `setup.py` and `requirements.txt`.
Then it is interesting to compare the performances (both computing time and model score) with our current implementation `optimize_params_bayes`. Could you do a test with the same number of optimisation iterations and 5-fold cross validation?

 I will have a look at the code in the next few days.

**Comment by luca-bottero** [comment]
Hi. I updated the requirements.
I have already performed a comparison between the performances of the bayesian method and the Optuna's one.
I firstly fixed the run time for Optuna at 10 minutes. The actual time was slightly longer than that by about one minute. Optuna actually did about 700 trials in that time. Then I fixed Bayes's number of iterations equal to the number of trials performed by Optuna. It took about 32 minutes to finish the task. In both cases the number of parallel threads was fixed.
In the following image I plotted the 5-fold ROC-AUC on the test set for each trial as well a linearly pondered Moving Average (MA, period = 50) of the two series. It can be seen that Optuna converges well after about 200 iterations, while Bayes does not shows such behavior.

opt_comp

In this particular case, the final (best) results of the optimization process are (5-fold ROC-AUC on test set):

1. 0.9949 - Optuna
2. 0.9947 - Bayes
3. 0.9942 - XGBoost's defaults parameters (for comparison)

**Comment by fmazzasc** [comment]
Hi @luca-bottero, 
thanks a lot for the optuna implementation. It would be nice to add the new method to the model handler test suite.  Could you do that? 
@fcatalan92 it looks to me that while the old bayes optimization method samples the hyper-params space uniformly (this is unexpected to me and should be understood) optuna improves the xgboost default nicely. I would merge the p.r. after adding the method to the tests.

**Comment by fcatalan92** [comment]
Hi @fmazzasc, indeed this needs to be investigated.

@luca-bottero Could you please share the code you used to do the check?

If the old optimisation method is not working, I think we can directly replace it with the optuna one after the checks. We can move the arguments for saving and continuing the optimisation of the optuna method in the `kwargs`, the API for the Bayesian optimisation would slightly change but it is fine IMHO.

**Comment by luca-bottero** [comment]
@fcatalan92 the code I used is in this repository https://github.com/luca-bottero/pp13TeV_hypertriton . I manually changed the config files in the main.py file after each phase described above. The plots are done in optimizer_analysis.py

**Comment by fcatalan92** [comment]
@luca-bottero Could you prepare a simple script that does the comparison?

Digging a bit in your repository, it looks to me that you kept the number of initialisation points `init_points` to its default when using `optimize_params_bayes`  https://github.com/luca-bottero/pp13TeV_hypertriton/blob/89eea3f2e1c686c3f778b31e2a73622d34d0aad9/src/utils.py#L98 . 
Could you try to increase them and look if is there any difference?
